generator/views/funnel_analysis_view.py

```diff
@@ -0,0 +1,238 @@
+
+        We only create a FunnelAnalysisView if we have the two necessary db tables:
+            - events_daily
+            - event_types
+        """
+        # We can guarantee there will always be at least one channel,
+        # because this comes from the associated _get_glean_repos in
+        # namespaces.py
+        dataset = next(
+            (channel for channel in channels if channel.get("channel") == "release"),
+            channels[0],
+        )["dataset"]
+
+        necessary_views = {"events_daily", "event_types"}
+        actual_views = {}
+        for view_id, references in db_views[dataset].items():
+            if view_id in necessary_views:
+                actual_views[view_id] = f"`mozdata.{dataset}.{view_id}`"
+
+        if len(actual_views) == 2:
+            # Only create an instance if we have the two necessary tables
+            tables = {
+                "funnel_analysis": "events_daily_table",
+                "event_types": actual_views["event_types"],
+            }
+            tables.update(
+                {
+                    f"event_type_{i}": "event_types"
+                    for i in range(1, num_funnel_steps + 1)
+                }
+            )
+            yield FunnelAnalysisView(
+                namespace,
+                [tables],
+            )
+
+    @classmethod
+    def from_dict(
+        klass, namespace: str, name: str, _dict: ViewDict
+    ) -> FunnelAnalysisView:
+        """Get a FunnalAnalysisView from a dict representation."""
+        return FunnelAnalysisView(namespace, _dict["tables"])
+
+    def to_lookml(self, bq_client, v1_name: Optional[str]) -> Dict[str, Any]:
+        """Get this view as LookML."""
+        return {
+            "includes": [f"{self.tables[0]['funnel_analysis']}.view.lkml"],
+            "views": self._funnel_analysis_lookml() + self._event_types_lookml(),
+        }
+
+    def n_events(self) -> int:
+        """Get the number of events allowed in this funnel."""
+        return len([k for k in self.tables[0] if k.startswith("event_type_")])
+
+    def _funnel_analysis_lookml(self) -> List[Dict[str, Any]]:
+        dimensions = [
+            {
+                "name": f"completed_event_{n}",
+                "type": "yesno",
+                "description": f"Whether the user completed step {n} on the associated day.",
+                "sql": (
+                    "REGEXP_CONTAINS(${TABLE}.events, mozfun.event_analysis.create_funnel_regex(["
+                    f"${{event_type_{n}.match_string}}],"
+                    "True))"
+                ),
+            }
+            for n in range(1, self.n_events() + 1)
+        ]
+        count_measures: List[Dict[str, Any]] = [
+            {
+                "name": f"count_user_days_event_{n}",
+                "description": (
+                    f"The number of user-days that completed step {n}. "
+                    "Grouping by day makes this is a count of users."
```

Comment:
This sentence doesn't quite parse for me.

Reply:
Okay, let me describe what this is and maybe you can help me uncover a better description.

For this explore, the analyst creates a funnel (step 1 -> step 2 -> ...). They then select this measure (e.g. `count_user_days_event_2`).

The underlying table has a grain of client-day: so for each client-day, we see if that funnel was completed. So if you simply selected this value, it would be the number of client-days that completed the funnel; however if you then selected `submission_date` as a dimension, each `submission_date` would give the number of clients who completed the funnel.

Reply:
How about this (hoping I understand your meaning):

`The total number of times that step {n} was completed. Grouping by day makes this a count of users who completed step {n} on each day.`

I would avoid using bespoke terminology like "client day" if at all possible.

Reply:
I like that, but then I have no idea what to call the measure. Maybe `count_completed_event_{n}`?

Reply:
I'm less worried about the name than the description but sounds good!